### ISSUE TITLE
The fallback on npmjs is not possible when at least one version of a module already exists in the local registry

### ISSUE THREAD

**Issue by Skywalker13** [hayes/unpm#75]
For example, I've a cache of many node modules in the uNPM local registry. But sometimes I've a module that depends of a newer version of a module which already exists in the cache. Then, the fallback on npmjs is ignored.

**Comment by hayes** [comment]
This is an interesting problem.  By design the npm client does not send the version of a dependency it is looking for, it just asks the server what version do you have and where can I get them.  There are several reasons for this, but the one I like the most is that that it is up to the client to figure out what version best matches the requirement in the `package.json` and you will always get the same semver semantics as long as you use the same version of the client (and adding to the semver spec like when they added `^`) does not require server side changes.

unpm can send you the public list if it does not have any version of the module you want, but if it has a version there is no way to tell if it will satisfy the dependency, so it can't know to send the fallback list of versions instead.

The only way I see around this is would be to always send a merge of the version un unpm and the versions in the fallback. This is problematic for a couple reasons.
 1) A large part of why some people use private registries is that they want to ensure consistent builds.
They get this because they know that the latest version in the registry works, and until they import something new, they wont get something new.  If you send down both the unpm and the fallback versions you lose this.

For example you depend on `foo@^1.1.1` and have `foo@1.1.2` in unpm and someone published  `foo@1.1.3` which despite semver breaks things in some way. suddenly you get a version from the fallback registry, despite have a working version in unpm.

2) every module name (even the private ones not published to the fallback) will be requested from npm, I know some people using unpm consider that a security issue.

So, long story short there are reasons that unpm works this way, but I can look into making an option that would send back both unpm and fallback versions on every request.  If you want to take a stab at PRing this in let me know and I'd be happy to give some guidance, otherwise I will try to find some to do this soon, but its I don't know how soon that might be.

**Comment by Skywalker13** [comment]
Thank you for this interesting answer.

IMHO, the first problematic reason is mostly not real because people which use the private registry for consistent builds only, keep the fallback disabled (and it's the default behavior).
The second point is interesting too. The problem can be a clash on the module's name. It is necessary to be sure that the private modules use a name which will be never used on the fallback registry.

For me, the first point is not a problem, and the second point is not more a problem because I use a namespace with my private modules.

---

I use uNPM for two reasons
1. I need a local registry for my private modules
2. I need a cache when I've no more Internet connection

Then, when I'm at work, I've Internet then it's fine even when I've only published my private modules. I publish the other modules (cache) in order to continue my work in the train (offline). When I go back at work, I must continue to work (the cache is not empty) and it's the problem.

Currently I've two options when I go back at work.
1. Delete the whole cache and re-publish only my private modules
2. Publish the new versions in the cache explicitly

The first way is acceptable
The second way is a bit annoying in my workflow... I can have many different versions of different modules and I can not spend too much time for that.

---

I've open this issue in order to have at least a trace somewhere of this "problem". I've no express necessity for an option for that. Maybe I will contribute to your project, but when; I don't know.
Prefer other features in your free time if you think that it is not an important point.

Thank you for this great project.

**Comment by hayes** [comment]
I have added `config.alwaysIncludeFallback` option in `1.1.0` hope this works for you, didn't have time to thoroughly test, but this should work.

**Comment by Skywalker13** [comment]
Great !
Thank you very much, I will test ASAP.

Regards

**Comment by Skywalker13** [comment]
I confirm that it works like a charm.

Thx
